pkg/pss/pss.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package pss
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/ethersphere/bee/pkg/tags"
+	"github.com/ethersphere/bee/pkg/trojan"
+)
+
+// Pss is the top-level struct, which takes care of message sending
+type Pss struct {
+	storer     storage.Storer
+	tags       *tags.Tags
+	handlers   map[trojan.Topic]Handler
+	handlersMu sync.RWMutex
+	//metrics metrics
+	//logger  logging.Logger
+}
+
+// Monitor is used for tracking status changes in sent trojan chunks
+type Monitor struct {
+	// returns the state of the trojan chunk that is being monitored
+	State chan tags.State
+}
+
+// NewPss inits the Pss struct with the localstore
+func NewPss(localStore storage.Storer, tags *tags.Tags) *Pss {
+	return &Pss{
+		storer:   localStore,
+		tags:     tags,
+		handlers: make(map[trojan.Topic]Handler),
+	}
+}
+
+// Handler defines code to be executed upon reception of a trojan message
+type Handler func(trojan.Message)
+
+// Send constructs a padded message with topic and payload,
+// wraps it in a trojan chunk such that one of the targets is a prefix of the chunk address
+// stores this in localstore for push-sync to pick up and deliver
+func (p *Pss) Send(ctx context.Context, targets trojan.Targets, topic trojan.Topic, payload []byte) (*Monitor, error) {
+	// TODO RESOLVE METRICS
+	//metrics.GetOrRegisterCounter("trojanchunk/send", nil).Inc(1)
+
+	//construct Trojan Chunk
+	m, err := trojan.NewMessage(topic, payload)
+	if err != nil {
+		return nil, err
+	}
+	var tc swarm.Chunk
+	tc, err = m.Wrap(targets)
+	if err != nil {
+		return nil, err
+	}
+
+	tag, err := p.tags.Create("pss-chunks-tag", 1, false)
+	if err != nil {
+		return nil, err
+	}
+
+	// SAVE trojanChunk to localstore, if it exists do nothing as it's already peristed
+	if _, err = p.storer.Put(ctx, storage.ModePutUpload, tc.WithTagID(tag.Uid)); err != nil {
+		return nil, err
+	}
+	tag.Total = 1
+
+	monitor := &Monitor{
+		State: make(chan tags.State, 3),
+	}
+
+	go monitor.updateState(tag)
+
+	return monitor, nil
+}
+
+// updateState sends the change of state thru the State channel
+// this is what enables monitoring the trojan chunk after it's sent
+func (m *Monitor) updateState(tag *tags.Tag) {
+	for _, state := range []tags.State{tags.StateStored, tags.StateSent, tags.StateSynced} {
+		for {
```

Comment:
this seems a bit un-go to me; wouldn't more go be count sends on a channel?

Reply:
This is a bit contrived, This should be changed, do any examples?, so I can make it more idiomatic. thanks

Reply:
At first glance it would require changes to the tags package, so it would emit on a channel when a count is increased. I don't know the package very well, though. Perhaps @acud @jmozah have better perspective.

Reply:
This was replaced with tags